### PR TITLE
Node.js vs. Django

### LINKED ISSUE #1096: Node.js vs. Django
### Brief Summary: 
Describe the what, why, and how of your content idea in 2-5 sentences.

Node.js and Django are technologies used in building versatile web and mobile applications. They have a large number of users, and both have a vast web community. They are both essential in developing web applications, and they are unique in their ways. 
The article will draw detailed comparisons of the two technologies, which will help the reader decide which one to use depending on the project needs. 

### Key Takeaways: 
What are the 3-5 most important things the reader should understand or be able to do after reading this article?

1.	In-depth comparisons between Node.js and Django. A comparison between the two in programming, architecture, security, templates and documentation, customization, scalability, performance, tools, etcetera. 
2.	Enumerate the advantages and disadvantages of using Node.js.
3.	Enumerate the advantages and disadvantages of using Django. 
4.	Types of apps that work well on each. 

### References: 
Please list links to any published content/research that you intend to use to support/guide this article. (If none, please indicate 'N/A'.)

### PR DESCRIPTION
**Pull request by Pius1991**
This fixes #1096

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process and style-once published, you can then submit any number of articles.)**
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

File: articles/node-js-vs-django/index.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
+Node.js and Django are technologies used in building versatile web and mobile applications. Both have an excellent user interface, are open, and do not require a license to operate, and therefore they have a vast number of users. Node.js is a framework that JavaScript uses on a client-side web application. Django works on [Python applications](https://www.javatpoint.com/python-applications) to help web developers build apps more quickly.
```
Comment (lalith1403) [inline articles/node-js-vs-django/index.md:1]: Django does not work on Python applications. It is used to build fully functional websites. Applications can be many. The link that you have given lists various applications possible with Python. Therefore, change the last sentence.

File: articles/node-js-vs-django/index.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+This article will analyze these two popular technologies. The article will critically give each member the pros and cons to help the reader choose the best technology depending on the project needs.
+
+### Node.js
+
+It is an open-source JavaScript run time. The first version was released in the [year 2009](https://nodejs.dev/learn/a-brief-history-of-nodejs). Node.js is based on the following programming languages; C, C++, and JavaScript. Node.js application is on the client-side browser. It provides scalability and concurrency.
```
Comment (lalith1403) [inline articles/node-js-vs-django/index.md:7]: Use the official node.js definition here.

File: articles/node-js-vs-django/index.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+#### Basic features of Node.js
```
Comment (lalith1403) [inline articles/node-js-vs-django/index.md:9]: Even Python has these features. Howcome JS is more prominent in building web applications? What separates Node.js from Python?

File: articles/node-js-vs-django/index.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+### Django
+
+It is an open-source Python web framework primarily designed to ensure high-speed web development. It is easy to learn for those knowledgeable in Python. Django has strict compliance with the rules and formalized scenarios for project implementation. Django provides accuracy in coding, a straight-forward app architecture, fewer errors, and bugs in the output.
+
+#### Basic features of Django
+
+1. Simplicity – Especially with the Python language, Django is easy to learn, use, and implement.
+2. It is an open-source language – No fees for installation and usage. It is available and freely accessible to the public.
+3. High-level language – It simplifies computer syntax into a language the user can easily understand.
```
Comment (lalith1403) [inline articles/node-js-vs-django/index.md:25]: Isn't JS also a high-level language?

File: articles/node-js-vs-django/index.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+4. It is both object-oriented and functional – meaning it can model real-world data and focus on functions. Node.js allows code re-useability.
```
Comment (lalith1403) [inline articles/node-js-vs-django/index.md:26]: reusability.

File: articles/node-js-vs-django/index.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+5. Portable – There is no need to write different codes for different machines as it is highly portable and flexible.
+6. It is available with a vast collection of libraries – Libraries are inbuilt, and therefore, there is no need to write unique code for everything.
```
Comment (lalith1403) [inline articles/node-js-vs-django/index.md:28]: Libraries are not inbuilt. But they are available. It is incorrect to say they are inbuilt. One needs to install libraries as and when required.

File: articles/node-js-vs-django/index.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+### Comparison between Django and Node.js
+
+| Distinctive point | Django | Node.js |
+| --- | --- | --- |
+| Programming language and primary use | It is a [Python based framework](https://www.upgrad.com/blog/python-frameworks/) used by developers to build computer apps within a stipulated time frame.| A framework that runs on JavaScript. Developers use it to build robust application programming interfaces (client-sided). |
```
Comment (lalith1403) [inline articles/node-js-vs-django/index.md:34]: JS is used for server-side applications as well.

File: articles/node-js-vs-django/index.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+| Architecture | Follows the [MTV model template](https://www.javatpoint.com/django-mvt) for purposes of handling data, validating, and interaction. | A runtime environment, therefore, works on an [event-driven model](https://www.redhat.com/en/topics/integration/what-is-event-driven-architecture). Runs on any operating system, maintaining a small number of requests. |
```
Comment (lalith1403) [inline articles/node-js-vs-django/index.md:35]: MVT model*

Comment (lalith1403) [reply articles/node-js-vs-django/index.md:35]: You need to explain the architecture. Do so in a separate section before you jump into the comparison.

File: articles/node-js-vs-django/index.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+| Security | It is more secure, comes with a built-in system, preventing a security breach. | It requires manual operations systems to manage security risks. |
+| Performance | It has a built-in house template system; hence the execution of tasks is quicker, thereby offering better performance. | It allows web professionals freedom in implementations. Though, it has a limitation of increasing the time required to build an application. |
+| Complexity | It is more complicated as a developer has to follow a particular specified path for solving problems. | It is less complicated as a developer has a free hand to solve problems. |
+| Efficiency | More efficient and faster, hence cost-effective. | It consumes more operation time; hence it is a less cost-effective option. |
```
Comment (lalith1403) [inline articles/node-js-vs-django/index.md:39]: This is debatable. For example, Customizability with Django is limited when compared to JS-based frameworks. Therefore, check this fact once.

File: articles/node-js-vs-django/index.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+#### Advantages
+
+- It is comparatively easy to learn.
+- Its built-in security system ensures security risks are very low.
+- It ensures app and web development is done at a fast rate and hence cost-effective.
+- The level of performance in the in-built templates is high.
+
+#### Disadvantages
+
+- It is not flexible, allows only [single-tiered](https://docs.bitnami.com/google-templates/singletier-vs-multitier/) applications.
```
Comment (lalith1403) [inline articles/node-js-vs-django/index.md:55]: Explain single-tiered applications.

File: articles/node-js-vs-django/index.md [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+- It has limitations with small apps. It does not work well in small apps.
```
Comment (lalith1403) [inline articles/node-js-vs-django/index.md:56]: Why?

File: articles/node-js-vs-django/index.md [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+- It requires much understanding of the framework.
```
Comment (lalith1403) [inline articles/node-js-vs-django/index.md:57]: much understanding -> detailed understanding. 
Also, isn't that the case with Node.js as well?

File: articles/node-js-vs-django/index.md [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+- It has many libraries around the world.
+- Features are not complicated; therefore, more comfortable to follow.
+- It is more flexible as developers have a free hand to execute commands.
+
+#### Disadvantages
+
+- It does not do well with intensive applications since it is a single thread system.
+- Less cost-effective, as it requires more time for a developer to develop a project.
+- It is prone to security flaws.
+- Its performance is low.
```
Comment (lalith1403) [inline articles/node-js-vs-django/index.md:73]: Why do you say so?

**Review by lalith1403 — CHANGES_REQUESTED** [review]

File: articles/node-js-vs-django/index.md (hunk 9 above)
Comment (Pius1991) [reply articles/node-js-vs-django/index.md:39]: > This is debatable. For example, Customizability with Django is limited when compared to JS-based frameworks. Therefore, check this fact once.

@lalith1403 please be specific on the issue, as I have not touched on customization with Django in the article.

Comment (lalith1403) [reply articles/node-js-vs-django/index.md:39]: I was referring to cost effectiveness. Cost effectiveness versus quality is my point.

Comment (Pius1991) [reply articles/node-js-vs-django/index.md:39]: @lalith1403 Thank you for your guidance, my argument is based on the fact that Django has in-built libraries and processes, a developer would not start from square one as it is for Node.js; hence less time, and therefore cost effective.

I understand your argument and I can see risk of misleading to the reader, let me just do away with that point.

File: articles/node-js-vs-django/index.md (hunk 2 above)
Comment (Pius1991) [reply articles/node-js-vs-django/index.md]: @lalith1403 I have included official definition

File: articles/node-js-vs-django/index.md (hunk 3 above)
Comment (Pius1991) [reply articles/node-js-vs-django/index.md:9]: > Even Python has these features. Howcome JS is more prominent in building web applications? What separates Node.js from Python?

I have made a new sentence that is making distinction.

File: articles/node-js-vs-django/index.md (hunk 4 above)
Comment (Pius1991) [reply articles/node-js-vs-django/index.md:25]: Yes it is, I have included that feature, thank you.

**Comment by Pius1991** [comment]
@lalith1403 I did the revisions as requested on the article, please check.

**Comment by hectorkambow** [comment]
plagiarism check done

**Review by louisefindlay23 — APPROVED** [review]
@hectorkambow, this article is ready for publication.

**Pull request merged**